types/react/index.d.ts

```diff
@@ -558,7 +629,37 @@ declare namespace React {
 
     function createRef<T>(): RefObject<T>;
 
-    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ComponentType<P & ClassAttributes<T>>;
+    interface ForwardRefExoticComponent<P> extends ExoticComponent<P> {
+        defaultProps?: Partial<P>;
+    }
+
+    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ForwardRefExoticComponent<P & ClassAttributes<T>>;
+
+    type ComponentProps<T extends ComponentType<any>> =
+        T extends ComponentType<infer P> ? P : {};
+    type ComponentPropsWithRef<T extends ComponentType<any>> =
+        T extends ComponentClass<infer P>
+            ? P & ClassAttributes<InstanceType<T>>
+            : T extends SFC<infer P>
+                ? P
+                : {};
+
+    interface MemoExoticComponent<T extends ComponentType<any>> extends ExoticComponent<ComponentPropsWithRef<T>> {
+        readonly type: T;
+    }
+
+    function memo<T extends ComponentType<any>>(
+        Component: T,
+        propsAreEqual?: (prevProps: Readonly<ComponentProps<T>>, nextProps: Readonly<ComponentProps<T>>) => boolean
+    ): MemoExoticComponent<T>;
+
+    interface LazyExoticComponent<T extends ComponentType<any>> extends ExoticComponent<ComponentPropsWithRef<T>> {
+        readonly _result: T;
```

Comment:
I'm not really happy about this. Having an underscored value here looks like reaching into internals; but it's the only way to have access to `T` in `LibraryManagedProps`.

An alternative is to make up our own fictional prop name.